Nodejs/Product/Npm/PackageSet.cs

```diff
@@ -0,0 +1,110 @@
+//*********************************************************//
+//    Copyright (c) Microsoft. All rights reserved.
+//    
+//    Apache 2.0 License
+//    
+//    You may obtain a copy of the License at
+//    http://www.apache.org/licenses/LICENSE-2.0
+//    
+//    Unless required by applicable law or agreed to in writing, software 
+//    distributed under the License is distributed on an "AS IS" BASIS, 
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or 
+//    implied. See the License for the specific language governing 
+//    permissions and limitations under the License.
+//
+//*********************************************************//
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.NodejsTools.Npm {
+    /// <summary>
+    /// Immutable collection of packages.
+    /// </summary>
+    public class PackageSet : IEnumerable<IPackage> {
```

Comment:
This logic can be simplified greatly. Instead of taking package diffs, let's just keep a list of package names that we've already tried downloading via tsd + the list of .d.ts files that we've already downloaded (it should be sufficient to query the package directory for any .d.ts files). Then simply compare this combined list against the names of the top level packages. This will be much simpler and less expensive. Again, we're not going for 100% correctness here - there's absolutely no way we're going to achieve that given the current state of DefinitelyTyped.